RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc

```diff
@@ -0,0 +1,1030 @@
+    if (d0CutOff_ > 0) {
+      Measurement1D atIP =
+	tk.stateAtBeamLine().transverseImpactParameter();// error contains beamspot
+      t_pi = 1. / (1. + local_exp(std::pow(atIP.value() / atIP.error(), 2) - std::pow(d0CutOff_, 2))); // reduce weight for high ip tracks
+      if (edm::isNotFinite(t_pi) ||  t_pi < std::numeric_limits<double>::epsilon())  continue; // usually is > 0.99
+    }
+    LogTrace("DAClusterizerinZT_vectorized") << t_z << ' ' << t_t <<' '<< t_dz2 << ' ' << t_dt2 <<' '<< t_pi;
+    tks.addItem(t_z, t_t, t_dz2, t_dt2, &tk, t_pi);
+  }
+  tks.extractRaw();
+  
+#ifdef VI_DEBUG
+  if (verbose_) {
+    std::cout << "Track count " << tks.getSize() << std::endl;
+  }
+#endif
+  
+  return tks;
+}
+
+
+namespace {
+  inline
+  double Eik(double t_z, double k_z, double t_dz2, double t_t, double k_t, double t_dt2) {
+    return std::pow(t_z - k_z, 2) * t_dz2 + std::pow(t_t - k_t,2) * t_dt2;
+  }
+}
+
+double DAClusterizerInZT_vect::update(double beta, track_t & gtracks,
+				     vertex_t & gvertices, bool useRho0, const double & rho0) const {
+
+  //update weights and vertex positions
+  // mass constrained annealing without noise
+  // returns the squared sum of changes of vertex positions
+  
+  const unsigned int nt = gtracks.getSize();
+  const unsigned int nv = gvertices.getSize();
+  
+  //initialize sums
+  double sumpi = 0.;
+  
+  // to return how much the prototype moved
+  double delta = 0.;
+  
+
+  // intial value of a sum
+  double Z_init = 0;
+  // independpent of loop
+  if ( useRho0 )
+    {
+      Z_init = rho0 * local_exp(-beta * dzCutOff_ * dzCutOff_); // cut-off
+    }
+  
+  // define kernels
+  auto kernel_calc_exp_arg = [ beta, nv ] ( const unsigned int itrack,
+					     track_t const& tracks,
+					     vertex_t const& vertices ) {
+    
+    const auto track_z = tracks.z_[itrack];
+    const auto track_t = tracks.t_[itrack];
+    const auto botrack_dz2 = -beta*tracks.dz2_[itrack];
+    const auto botrack_dt2 = -beta*tracks.dt2_[itrack];
+
+    // auto-vectorized
+    for ( unsigned int ivertex = 0; ivertex < nv; ++ivertex) {
+      const auto mult_resz = track_z - vertices.z_[ivertex];
+      const auto mult_rest = track_t - vertices.t_[ivertex];
+      vertices.ei_cache_[ivertex] = botrack_dz2 * ( mult_resz * mult_resz ) + botrack_dt2 * ( mult_rest * mult_rest );
+    }
+  };
+  
+  auto kernel_add_Z = [ nv, Z_init ] (vertex_t const& vertices) -> double
+    {
+      double ZTemp = Z_init;
+      for (unsigned int ivertex = 0; ivertex < nv; ++ivertex) {	
+	ZTemp += vertices.pk_[ivertex] * vertices.ei_[ivertex];
+      }
+      return ZTemp;
+    };
+
+  auto kernel_calc_normalization = [ beta, nv ] (const unsigned int track_num,
+						  track_t & tks_vec,
+						  vertex_t & y_vec ) {
+    auto tmp_trk_pi = tks_vec.pi_[track_num];
+    auto o_trk_Z_sum = 1./tks_vec.Z_sum_[track_num];
+    //auto o_trk_dz2 = tks_vec.dz2_[track_num];
+    //auto o_trk_dt2 = tks_vec.dt2_[track_num];
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant